File: cpp/139. Word Break.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef vector<int> vi;
typedef vector<vector<int > > vvi;
typedef pair<int, int> pi;

/*
dp[i]: True if s[i...n-1] is word breakable
dp[i] = any(s[i...j-1] in wordDict && dp[j] for j in [i+1, n])
dp[n] = true
ans = dp[0]
*/

class Solution {
// DP, Time: O(n^2), Space: O(n)
public:
    bool wordBreak(string s, vector<string>& wordDict) {
        unordered_set<string> dict(wordDict.begin(), wordDict.end());       
        int n = s.size();
        bool dp[n+1]; fill(dp, dp+n+1, false); dp[n] = true;
        for(int i = n-1; i >=0; i--) {
            for(int j = i+1; j <= n; j++) {
                string tmp = s.substr(i, j-i);
                if(dp[j] && dict.count(tmp)) {
                    dp[i] = true;
                    break;
                }
            }
        }
        return dp[0];
    }
};

int main() {
    ios::sync_with_stdio(0); 
    cin.tie(0);

    Solution solution; string s; vector<string> wordDict; bool ans;
    
    solution = Solution();
    s = "leetcode";
    wordDict = {"leet", "code"};
    ans = solution.wordBreak(s, wordDict);
    assert(ans==true);

    solution = Solution();
    s = "applepenapple";
    wordDict = {"apple", "pen"};
    ans = solution.wordBreak(s, wordDict);
    assert(ans==true);

    solution = Solution();
    s = "catsandog";
    wordDict = {"cats", "dog", "sand", "and", "cat"};
    ans = solution.wordBreak(s, wordDict);
    assert(ans==false);
}

File: python/139-Word Break.py
from typing import List
from functools import lru_cache

'''
https://www.youtube.com/watch?v=YxtQUbKbdUs
Be careful that dp[i] is for s[0, i) not include i!!!!!
dp[i]: True if s[0, i) is word breakable, i=[1, len(s))
dp[0]: True, for empty string case

dp[i] == True, if there exists an j= [0, i) such that
1. s[0, j) is word breakable (i.e. dp[j] == True) and
2. s[j, i) is in wordDict

return dp[-1]

ex: leetcode, [leet, code]
         l  e  e  t  c  o  d  e
dp = [T, F, F, F, T, F, F ,F, T]
'''

class Solution:
    def wordBreak(self, s: str, wordDict) -> bool:
        "DFS + Memorization"
        wordDict = set(wordDict)
        n = len(s)
        @lru_cache(None)
        def dfs(i):
            if i == n:
                return True
            for j in range(i+1, n+1):
                if s[i:j] in wordDict:
                    if dfs(j): return True
            return False
        return dfs(0)
    def wordBreak(self, s: str, wordDict) -> bool:
        "Down Top DP, Time: O(m + n^2), Space: O(n+m), n: len(s), m: len(wordDict)"
        n = len(s)
        if len(s) == 0: return False
        wordDict = set(wordDict)
        dp = [False]*(n+1) # dp[i]: [0, i)
        dp[0] = True
        for i in range(1, n+1):
            for j in range(i):
                if (dp[j] and s[j:i] in wordDict):
                    dp[i] = True
                    break
        return dp[n]

s = "leetcode"
wordDict = ["leet", "code"]
result = Solution().wordBreak(s, wordDict)
assert result == True

s = "applepenapple"
wordDict = ["apple", "pen"]
result = Solution().wordBreak(s, wordDict)
assert result == True


s = "catsandog"
wordDict = ["cats", "dog", "sand", "and", "cat"]
result = Solution().wordBreak(s, wordDict)
assert result == False
    